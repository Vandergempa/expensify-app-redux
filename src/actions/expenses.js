import uuid from 'uuid';
import database from '../firebase/firebase';
// action generators!
// export const addExpense = (
//     { 
//         description = '', 
//         note = '', 
//         amount = 0,
//         createdAt = 0 
//    } = {}
// ) => ({
//    type: 'ADD_EXPENSE',
//    expense: {
//        id: uuid(),
//        description,
//        note,
//        amount,
//        createdAt
//    }
// });

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

// Because we set up the middleware for redux thunk, now it is possible to return a function instead of
// an object:
export const startAddExpense = (expenseData = {}) => {
   // In this case we are going to return a function. The function gets called with dispatch.
   // This lets us use dispatch in the function.
   return (dispatch) => {
      // This method of defining default is the same as above:
      const {
         description = '', 
         note = '', 
         amount = 0,
         createdAt = 0 
      } = expenseData;
      const expense = { description, note, amount, createdAt };

      return database.ref('expenses').push(expense).then((ref) => {
         // THis dispatch is gonna make sure that the redux store changes too.
         dispatch(addExpense({
            id: ref.key,
            ...expense
         }));
      });
   };
};

// REMOVE_EXPENSE

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id }) => {
   return (dispatch) => {
      return database.ref(`expenses/${id}`).remove().then(() => {
         dispatch(removeExpense({id}));
      });
   };
};

// EDIT_EXPENSE

export const editExpense = (id, updates) => ({
   type: 'EDIT_EXPENSE',
   id,
   updates
});

// SET_EXPENSES
export const setExpenses = (expenses) => ({
   type: 'SET_EXPENSES',
   expenses
});

export const startSetExpenses = () => {

   return (dispatch) => {
      return database.ref("expenses")
      .once('value')
      .then((snapshot) => {
        const expenses = [];
    
        snapshot.forEach((childSnapshot) => {
            expenses.push({
                id: childSnapshot.key,
                ...childSnapshot.val()
            });
        });
        console.log(expenses); 
        dispatch(setExpenses(expenses));  
      });
   };
};
