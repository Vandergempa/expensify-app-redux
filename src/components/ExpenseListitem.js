import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

export const ExpenseListitem = ({ dispatch, id, description, amount, createdAt }) => (
    // {dispatch} was needed for the button which was later moved to the editexpensepage
    <div>
        <Link to={`/edit/${ id }`} >
            <h3>{ description }</h3>
        </Link>
        <p> 
            { numeral(amount / 100).format('$0.00') } 
            - 
            { moment(createdAt).format('do MMMM, YYYY') }
        </p>
    </div>
);

export default ExpenseListitem;