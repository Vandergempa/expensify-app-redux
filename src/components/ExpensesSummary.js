import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import getExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
    const expenseOrExpenses = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0.00');
        return (
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Viewing <span>{expenseCount}</span> {expenseOrExpenses} totalling <span>{formattedExpensesTotal}</span>.</h1>
                    <div className="page-header__actions">
                        <Link className="button" to="/create">Add Expense</Link>
                    </div>
                </div>
            </div>
        )
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    }
};

export default connect(mapStateToProps)(ExpensesSummary);