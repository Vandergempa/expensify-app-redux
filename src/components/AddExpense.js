import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

// to export the unconnected component too
export class AddExpense extends React.Component {
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense));
        //* props.dispatch is changed so that testing can be done... as a result mapDispatchToProps
        //* is created with the dispatch logic
        this.props.startAddExpense(expense);
        //switching to homepage using browser routing
        this.props.history.push('/');
    };
    render() {
        return (
    <div>
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Add expense</h1>
            </div>
        </div>
        <div className="content-container">
            <ExpenseForm 
                buttonText= "Add Expense"
                onSubmit={this.onSubmit}
            />
        </div>
    </div>
        )
    }
}

// It is a way to return the dispatcher functions allowing to obsctruct away from the component itself.
const mapDispatchToProps = (dispatch) => {
    return {
        startAddExpense: (expense) => dispatch(startAddExpense(expense))
    };
};

export default connect(undefined, mapDispatchToProps)(AddExpense);