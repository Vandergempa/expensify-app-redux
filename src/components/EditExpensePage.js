import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense, startRemoveExpense, startEditExpense } from '../actions/expenses';


export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
         // props.dispatch(editExpense(props.expense.id, expense));
        //* props.dispatch is changed so that testing can be done... as a result mapDispatchToProps
        //* is created with the dispatch logic
        this.props.startEditExpense(this.props.expense.id, expense);
        //switching to homepage using browser routing
        this.props.history.push('/');    
    };
    onRemove = () => {
       this.props.startRemoveExpense({ id: this.props.expense.id });
       this.props.history.push('/');    
   };

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm                    
                        buttonText= "Edit Expense"
                        expense= {this.props.expense}
                        onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary" onClick= {this.onRemove} >Remove Expense</button>
                </div>
            </div>
        )
    }
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id;
        })
    };
};

// It is a way to return the dispatcher functions allowing to obsctruct away from the component itself.
const mapDispatchToProps = (dispatch, props) => {
    return {
        startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
        startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
