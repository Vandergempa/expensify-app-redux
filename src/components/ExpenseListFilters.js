import React from 'react';
import { connect } from 'react-redux';
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';


export class ExpenseListFilters extends React.Component {
    state = {
        focusedInput: null,
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    onFocusChange = ( focusedInput ) => {
        this.setState(() => ({ focusedInput: focusedInput }))
    };
    onTextChange = (e) => {
            this.props.setTextFilter(e.target.value);
    };
    onSortChange = (e) => {
        if (e.target.value === "date") {
            this.props.sortByDate();
        } else if (e.target.value === "amount") {
            this.props.sortByAmount();
        }
    }
    render() {
        return(
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange}
                />
                <select 
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange} 
                >
                <option value="date">Date</option> 
                <option value="amount">Amount</option> 
                </select>
                    <DateRangePicker 
                        startDate={this.props.filters.startDate}
                        startDateId="your_unique_start_date_id"
                        endDate={this.props.filters.endDate}
                        endDateId="your_unique_end_date_id"
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.focusedInput}
                        onFocusChange={this.onFocusChange}
                        showClearDates={true}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
            </div>
        );
    }
};


const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

// It is a way to return the dispatcher functions allowing to obsctruct away from the component itself.
const mapDispatchToProps = (dispatch, props) => {
    return {
       setTextFilter: (text) => dispatch(setTextFilter(text)),
       sortByDate: () => dispatch(sortByDate()),
       sortByAmount: () => dispatch(sortByAmount()),
       setStartDate: (startDate) => dispatch(setStartDate(startDate)),
       setEndDate: (endDate) => dispatch(setEndDate(endDate))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);