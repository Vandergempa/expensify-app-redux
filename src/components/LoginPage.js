import React from 'react';
import { connect } from 'react-redux';
import { startLoginGoogle, startLoginFacebook } from '../actions/auth';

// Destructuring startLogin from props(mapDispatchToProps from the store):
export const LoginPage = ({ startLoginGoogle, startLoginFacebook }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title" >Expensify</h1>
            <h3>It's time to get your expenses under control.</h3>
            <p>Choose to login:</p>
            <a className="button__login" onClick={startLoginFacebook}><img src="images/facebook_squared.png"></img></a>
            <a className="button__login" onClick={startLoginGoogle}><img src="images/google_icon.png"></img></a>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLoginGoogle: () => dispatch(startLoginGoogle()),
    startLoginFacebook: () => dispatch(startLoginFacebook())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
