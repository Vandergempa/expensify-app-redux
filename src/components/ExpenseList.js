import React from 'react';
import { connect } from 'react-redux';
import ExpenseListitem from './ExpenseListitem';
import selectExpenses from '../selectors/expenses';

// we are exporting it for testing
export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>There are no expenses to show</span>
                    </div>
                ) : (
                    props.expenses.map((expense) => {
                        return <ExpenseListitem key={expense.id} {...expense}/>;
        // Or spreading can be used {...expense}, then ExpenseListitem can be written without using {expense}
                    })
                )
            }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters),
    };
}

export default connect(mapStateToProps)(ExpenseList);
