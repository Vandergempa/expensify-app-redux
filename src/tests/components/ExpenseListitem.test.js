import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseListitem } from '../../components/ExpenseListitem';
import expenses from '../fixtures/expenses';

test('Should render a single ExpenseListitem', () => {
    const wrapper = shallow(<ExpenseListitem {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot();
});