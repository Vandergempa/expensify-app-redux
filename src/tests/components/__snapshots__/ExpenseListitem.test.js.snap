// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render a single ExpenseListitem 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
      Gum
    </h3>
  </Link>
  <p>
    $2.00
    -
    4th January, 1970
  </p>
</div>
`;
